### PR TITLE
rolling_exp: keep_attrs and typing

### PR DESCRIPTION
**Pull request by mathause**
- [x] Towards #4513
 - [x] Tests added
 - [x] Passes `isort . && black . && mypy . && flake8`
 - [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

File: doc/whats-new.rst [hunk 1]
```diff
@@ -45,6 +45,10 @@ New Features
   By `Michal Baumgartner <https://github.com/m1so>`_.
 - :py:meth:`Dataset.weighted` and :py:meth:`DataArray.weighted` are now executing value checks lazily if weights are provided as dask arrays (:issue:`4541`, :pull:`4559`).
   By `Julius Busecke <https://github.com/jbusecke>`_.
+- Added the ``keep_attrs`` keyword to :py:meth:`~xarray.DataArray.rolling_exp.mean` and :py:meth:`~xarray.Dataset.rolling_exp.mean`.
+  The attributes are now kept per default.
```
Comment (mathause) [inline doc/whats-new.rst:49]:
```suggestion
  The attributes are now kept per default (:pull:`4592`).
```

File: xarray/core/rolling_exp.py [hunk 2]
```diff
@@ -1,8 +1,17 @@
...
 
+from .options import _get_keep_attrs
 from .pdcompat import count_not_none
 from .pycompat import is_duck_dask_array
 
+if TYPE_CHECKING:
+    from .dataarray import DataArray  # noqa: F401
+    from .dataset import Dataset  # noqa: F401
+
+T_DSorDA = TypeVar("T_DSorDA", "DataArray", "Dataset")
```
Comment (max-sixty) [inline xarray/core/rolling_exp.py:13]: Is this different from `DataWithCoords`?

File: xarray/core/rolling_exp.py [hunk 3]
```diff
@@ -56,7 +65,7 @@ def _get_center_of_mass(comass, span, halflife, alpha):
     return float(comass)
 
 
-class RollingExp:
+class RollingExp(Generic[T_DSorDA]):
```
Comment (max-sixty) [inline xarray/core/rolling_exp.py:68]: Great idea with `Generic`

**Review by max-sixty — APPROVED** [review]
Thanks @mathause ! Excellent tests!

More generally — if `keep_attrs` doesn't already exist as a kwarg in functions, should we add it? Or just be true?

(tbc, either way having the default be true here is a very welcome feature)

File: xarray/core/rolling_exp.py (hunk 3 above)
Comment (mathause) [reply xarray/core/rolling_exp.py:68]: Yes that took quite a while to figure out but it doesn't work without it as the Type of `obj` does not bind to the function else. That should come in handy for some other classes.

File: xarray/core/rolling_exp.py (hunk 2 above)
Comment (mathause) [reply xarray/core/rolling_exp.py:13]:
```python
from .common import DataWithCoords
T_DataWithCoords = TypeVar("T_DataWithCoords", bound=DataWithCoords)
```
works as well. A bit to my surprise as `DataWithCoords` does not implement `reduce`. But mypy does not throw an error here: 

```python
from xarray.core.common import DataWithCoords
class A:
    pass

hasattr(xr.core.common.DataWithCoords, "reduce") # -> False

def test(x: "A"):
    x.reduce() # mypy errors

def test2(x: "DataWithCoords"):
    x.reduce() # mypy does not error
```
not sure why not...

---

(Also I am not entirely sure what the difference is between
```python
T_Parent = TypeVar("T_Parent", bound=Parent)
T_Children = TypeVar("T_Children", ChildA, ChildB)
```
)

**Comment by mathause** [comment]
> More generally — if keep_attrs doesn't already exist as a kwarg in functions, should we add it? Or just be true?

I think adding a `kwarg` does not hurt much but it's probably not used often, either. I don't have a strong opinion.

File: xarray/core/rolling_exp.py (hunk 2 above)
Comment (max-sixty) [reply xarray/core/rolling_exp.py:13]: That is odd it doesn't throw an error... 

On the second case, I think — not completely sure — that `T_Parent` is generic with an "upper bound" of `Parent`, so any function that takes `T_Parent` will return the same type. 

Whereas a function that takes `T_Children` could return either of `ChildA` or `ChildB` regardless of what's passed in.

Comment (max-sixty) [reply xarray/core/rolling_exp.py:13]: And your call on whether we use `DataWithCoords` or the `TypeVar`, but I think if we do the latter we could find a clearer name :)

I would marginally vote to go with `DataWithCoords` given we do that elsewhere but totally open on trying another approach

Comment (mathause) [reply xarray/core/rolling_exp.py:13]: I think both would need to be `TypeVar`s. I I'll stick to `T_DSorDA` as `DataWithCoords` does not implement `reduce`. `T_DSorDA` is already used at some places (`map_blocks`), e.g.:

https://github.com/pydata/xarray/blob/5296ed18272a856d478fbbb3d3253205508d1c2d/xarray/core/parallel.py#L34

Comment (max-sixty) [reply xarray/core/rolling_exp.py:13]: Ah I didn't know that, thanks. I wasn't a fan of the name, but we can change them together another time if others agree.

**Comment by max-sixty** [comment]
Thanks @mathause !

**Pull request merged**